docs/api/ref/GetAgent.md

```diff
@@ -0,0 +1,125 @@
+
+```Typescript
+/** 
+ * Retrieves an FDC3 DesktopAgent instance either from a Desktop Agent that supports
+ * injection or by using the FDC3 Web Connection Protocol (WCP) to establish a connection
+ * with a browser-resident Desktop Agent.
+ * 
+ * @param {GetAgentParams} params Required parameters for the function.
+ *  
+ * @return A promise that resolves to an object which contains a DesktopAgent, or which
+ * rejects with a string from the AgentError union if a DesktopAgent cannot be established.
+ * 
+ * @example  
+ * const { desktopAgent: fdc3 } = await getAgent({ 
+ *     appId: “myApp@myorg.com/appd”
+ * }); 
+ * 
+ *  * @example Using appDUrl
+ * const { desktopAgent: fdc3 } = await getAgent({ 
+ *     appId: “myApp”,
+ *     appDUrl: "myorg.com/appd"
+ * }); 
+ */
+
+export type GetAgentFunction = ( 
+	params: GetAgentParams,  
+) => Promise<{
+    desktopAgent: DesktopAgent,
+    implementationMetadata: ImplementationMetadata,
+    appMetadata: AppMetadata
+}>;
+
+/** 
+ * @typedef {Object} GetAgentParams Type representing parameters passed to the 
+ * getAgent function.
+ * 
+ * @property {string} appId The fully qualified appId that represents the application.
```

Comment:
Yes it happens on same domain window navigations (we work with many firms that publish many apps, both in-house and clients and often use a common domain), but also where firms use certain types of proxy that replace domains and SSL certificates for external sites. 

The latter seems a weird practice to me, but we've seen it at a few firms. The former is more often going to be the issue as its common for those apps to be able to navigate to each other with a built-in nav menu.

Further, this is a key part of apps self identifying - it's the link back to the app directory that contains their identity (which as discussed may not be the one they were launched with - but its fundamental that the app code needs to contain a link to the identity, rather than a 3rd party directory). This has been discussed at length in the meetings on several occasions.

Reply:
Resolving as we have to have self identification of apps to resolve appIds wherever there are multiple apps running on the same domain AND to confirm that they did not change identity on navigation (they could navigate to another related app or a user could simply type in a different address of path in the address bar)

Reply:
I've been thinking about this a lot in the intervening time.    

Perhaps we need to "bite the bullet" here and add the app's public key to the directory?   Then, the DA can issue a challenge ("hey, encrypt this token with your private key") and the app can respond with something the DA can check.  

This is how symphony works, pretty much.

Reply:
In any case I think we need a better term than `instanceUuid`.  This is too close to `instanceId` and doesn't really reflect that it's something unique to the app rather than the app instance.

Reply:
> In any case I think we need a better term than instanceUuid. This is too close to instanceId and doesn't really reflect that it's something unique to the app rather than the app instance.

Re: `instanceUuid`, its not used for app identity but instance identity. It should be unique for each `instanceId`. It's a shared secret generated by the DA when it sees a new instance to allow that instance to get the same instanceId again and should only be known to the DA and that particular instance (only that instance can access the SessionStorage we put it into - different windows for the same app get separate SessionStorage).

It's needed as, if you have two instances of the same app, and they navigate or refresh it's hard for the DA to tell which is which. As per the proposal doc there are edge cases to comparing the window proxy objects (if the app navigate the window proxy may still be the same), there also edges in the `instanceUuid` approach (`window.open` to same domain can clone the `SessionStorage`) so the validation procedure for **instance identity** is to compare both the `instanceUuid` and the `WindowProxy` reference in the postMessage to the existing one you have, both having to match for it to be issued the same `instanceId`. We shouldn't use the `instanceId` itself for this as that's not a secret/available to other apps through the API in various ways.

Re: Biting the bullet and adding a public key to the appD - absolutely! However, as per your previous comment we didn't want to step on that group's toes and stated we'd adopt whatever identity + SWG adopts for FDC3. Pretty sure that'll be a key part of it so they could always be raising that issue and PR ASAP. I was expecting that something for both topics would make it in to FDC3 2.2 and hence we'd adjust here as soon as we/they had something there ;-)